plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt

```diff
@@ -0,0 +1,177 @@
+ */
+private fun kotlinClassHeaderVisitor(body: (KotlinClassHeader) -> Unit): AnnotationVisitor =
+    object : AnnotationVisitor(Opcodes.API_VERSION) {
+        var kind: Int = 1
+        var metadataVersion: IntArray = intArrayOf()
+        var bytecodeVersion: IntArray = intArrayOf()
+        var data1: MutableList<String> = mutableListOf()
+        var data2: MutableList<String> = mutableListOf()
+        var extraString: String = ""
+        var packageName: String = ""
+        var extraInt: Int = 0
+
+        override fun visit(name: String, value: Any?) {
+            when (name) {
+                "k" -> kind = value as Int
+                "extraInt" -> extraInt = value as Int
+                "mv" -> metadataVersion = value as IntArray
+                "bv" -> bytecodeVersion = value as IntArray
+                "xs" -> extraString = value as String
+                "pn" -> packageName = value as String
+            }
+        }
+
+        override fun visitArray(name: String): AnnotationVisitor {
+            val destination = when (name) {
+                "d1" -> data1
+                "d2" -> data2
+                else -> error("Wrong array field in kotlin metadata: $name")
+            }
+            return object : AnnotationVisitor(Opcodes.ASM7) {
+                override fun visit(name: String?, value: Any?) {
+                    destination += value as String
+                }
+            }
+        }
+
+
+        override fun visitEnd() {
+            body(KotlinClassHeader(
+                kind,
+                metadataVersion,
+                bytecodeVersion,
+                data1.toTypedArray(),
+                data2.toTypedArray(),
+                extraString,
+                packageName,
+                extraInt
+            ))
+        }
+    }
+
+/**
+ * Serialize a KotlinClassHeader to an existing Kotlin Metadata annotation visitor.
+ */
+private fun AnnotationVisitor.visitKotlinMetadata(header: KotlinClassHeader) {
+    visit(JvmAnnotationNames.KIND_FIELD_NAME, header.kind)
+    visit(JvmAnnotationNames.METADATA_VERSION_FIELD_NAME, header.metadataVersion)
+    visit(JvmAnnotationNames.BYTECODE_VERSION_FIELD_NAME, header.bytecodeVersion)
+    visitArray(JvmAnnotationNames.METADATA_DATA_FIELD_NAME).apply {
+        header.data1.forEach { visit(null, it) }
+        visitEnd()
+    }
+    visitArray(JvmAnnotationNames.METADATA_STRINGS_FIELD_NAME).apply {
+        header.data2.forEach { visit(null, it) }
+        visitEnd()
+    }
+    visit(JvmAnnotationNames.METADATA_EXTRA_STRING_FIELD_NAME, header.extraString)
+    visit(JvmAnnotationNames.METADATA_PACKAGE_NAME_FIELD_NAME, header.packageName)
+    visit(JvmAnnotationNames.METADATA_EXTRA_INT_FIELD_NAME, header.extraInt)
+    visitEnd()
+}
+
+/**
+ * Class metadata adapter which removes private functions, properties, type aliases,
+ * and local delegated properties.
+ */
+private class AbiKmClassVisitor(delegate: KmClassVisitor) : KmClassVisitor(delegate) {
```

Comment:
Shouldn't we also remove private nested classes, and private constructors?

Reply:
Yes, absolutely. Fixed and also removed private Companion objects.

I guess we should not remove private sealed subclasses due to exhaustiveness checking, though?

Reply:
Oh. I wonder if removing the nested classes and keeping the references in sealed subclasses breaks the exhaustiveness checking anyway, because the compiler can't load the nested class correctly and thus may skip loading it altogether (or worse, crash with some exception).

Reply:
That part is not an issue right now, since we only strip nested anonymous classes.

This is necessary for Java, since the classes will still be package visible, but I wonder if we should strip more classes after all. In principle this could break Java code, though, so I'm not completely convinced. What do you think?

Reply:
Sorry for the delay. I'm probably missing something. My initial comment was about private nested classes, and we now do strip them, after a new change you've added.

However, now I worry that that might have been risky, because the Kotlin compiler treats the "nested classes" field as the only source of truth. A class named `A$B` _possibly_ won't be treated as a nested class named `B` inside `A`, unless `A` has `B` in its nested classes in the metadata. Whether that would lead to any noticeable problems is an open question though.

In case of private sealed subclasses which are nested, the metadata becomes _sort of_ inconsistent now -- the "sealed subclasses" field mentions some classes which are clearly nested in their outer class, yet that outer class knows nothing about them, because they are absent in its "nested classes" field in the metadata. Again, I'm not sure that it would cause any problems.

From the data organization point of view, all nested classes should be declared in the "nested classes" field. Any other data, including "companion object" and "sealed subclasses", does not _declare_ new classes, but only references the existing ones declared in "nested classes".

Reply:
On reflection, the current code seems to be completely useless. Right now we only strip (some) local anonymous classes. In particular, none of the classes that we remove would end up in the Kotlin metadata in the first place. We might want to revisit this later, since there are definitely more classes that we could strip, but for now I've removed the code in question.